### PR TITLE
add tab extension to dev console

### PR DESCRIPTION
**Pull request by jgbernalp**
This PR adds a new extension type to be able to add tabs on the `HosrizontalNav` component in specific sections 

Screenshot 2022-09-28 at 11 33 16

https://issues.redhat.com/browse/OU-49

### CODE REVIEW

File: frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md [hunk 1]
```diff
@@ -972,6 +973,22 @@ Adds a new storage class provisioner as an option during storage class creation.
...
+### Summary 
+
+Contributes a tab into an HorizontalNav component matching the sectionId.
+
+### Properties
+
+| Name | Value Type | Optional | Description |
+| ---- | ---------- | -------- | ----------- |
+| `contextId` | `string` | no | ContextId assigned to the horizontal nav in which the tab will be injected |
+| `page` | `{ name: string; href: string; }` | no | The page to be show in horizontal tab. It takes tab name as name and href of the tab |
```
Comment (spadgett) [inline frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md:987]: We should clarify what `name` is. Is this the label of the tab or some internal name?

File: frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md [hunk 2]
```diff
@@ -972,6 +973,22 @@ Adds a new storage class provisioner as an option during storage class creation.
 
 ---
 
+## `console.tab`
+
+### Summary 
+
+Contributes a tab into an HorizontalNav component matching the sectionId.
```
Comment (spadgett) [inline frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md:980]:
```suggestion
Adds a tab to an `HorizontalNav` matching the `contextId`.
```

Comment (spadgett) [reply frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md:980]: A code example would be good here.

Comment (spadgett) [reply frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md:980]: We should discuss how ordering works.

File: frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md [hunk 3]
```diff
@@ -972,6 +973,22 @@ Adds a new storage class provisioner as an option during storage class creation.
...
+
+### Properties
+
+| Name | Value Type | Optional | Description |
+| ---- | ---------- | -------- | ----------- |
+| `contextId` | `string` | no | ContextId assigned to the horizontal nav in which the tab will be injected |
```
Comment (spadgett) [inline frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md:986]:
```suggestion
| `contextId` | `string` | no | Context ID assigned to the horizontal nav in which the tab will be injected |
```

Comment (spadgett) [reply frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md:986]: Should we list the known context IDs?

File: frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md (hunk 1 above)
Comment (spadgett) [reply frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md:987]: When we use `href` here, what do we mean? Is it just the path appended to the existing URL? (If so, I'm not sure `href` is the right property name, but we probably use that elsewhere.)

File: frontend/public/components/utils/horizontal-nav.tsx [hunk 4]
```diff
@@ -276,22 +282,38 @@ export const HorizontalNav = React.memo((props: HorizontalNavProps) => {
...
+    const resolvedNavTab = navExtentionsResolved
+      ? dynamicTabExtensions
+          .filter((tab) => tab.properties.contextId === contextId)
+          .map((tab) => ({
+            ...tab.properties.page,
+            component: tab.properties.component,
+          }))
+      : [];
+
+    return [...resolvedResourceNavTab, ...resolvedNavTab];
```
Comment (spadgett) [inline frontend/public/components/utils/horizontal-nav.tsx:308]: We need to ensure there is stable ordering if multiple plugins extend the same horizontal nav.

File: frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md (hunk 2 above)
Comment (christianvogt) [reply frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md:980]: `a HorizontalNav`

File: frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md (hunk 3 above)
Comment (christianvogt) [reply frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md:986]: Inconsistencies in docs with `HorizontalNav` and `horizontal nav`

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/horizontal-nav-tabs.ts [hunk 5]
```diff
@@ -28,6 +28,23 @@ export type HorizontalNavTab = ExtensionDeclaration<
...
+export type NavTab = ExtensionDeclaration<
+  'console.tab',
+  {
+    /** ContextId assigned to the horizontal nav in which the tab will be injected */
+    contextId: string;
+    /** The page to be show in horizontal tab. It takes tab name as name and href of the tab */
+    page: {
+      name: string;
+      href: string;
+    };
```
Comment (christianvogt) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/horizontal-nav-tabs.ts:41]: How did we ever agree that this was the correct definition for the resource tabs?
There's no reason for nesting these properties under `page`.

@spadgett should we keep consistency with existing resource tab extension or just go ahead and make improvements now?

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/horizontal-nav-tabs.ts [hunk 6]
```diff
@@ -28,6 +28,23 @@ export type HorizontalNavTab = ExtensionDeclaration<
...
+    /** Tab content component. */
+    component: CodeRef<React.ComponentType<PageComponentProps>>;
```
Comment (christianvogt) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/horizontal-nav-tabs.ts:43]: Is `PageComponentProps` correct here? TBH I don't know what some of the props do for the horizontal nav tabs extension.

File: frontend/public/components/utils/horizontal-nav.tsx [hunk 7]
```diff
@@ -258,9 +260,13 @@ export const HorizontalNav = React.memo((props: HorizontalNavProps) => {
...
-  const [dynamicNavTabExtensions, navTabExtentionsResolved] = useResolvedExtensions<
-    DynamicHorizontalNavTab
-  >(DynamicIsHorizontalNavTab);
+  const contextId = props.contextId;
+  const [dynamicResourceNavTabExtensions, navTabResourceExtentionsResolved] = useResolvedExtensions<
+    DynamicResourceNavTab
+  >(DynamicIsResourceNavTab);
+  const [dynamicTabExtensions, navExtentionsResolved] = useResolvedExtensions<DynamicNavTab>(
+    DynamicIsNavTab,
+  );
```
Comment (christianvogt) [inline frontend/public/components/utils/horizontal-nav.tsx:269]: Looks like we've been doing it wrong this whole time.
It's incorrect to resolve the extensions here. We don't want to resolve code references that aren't relevant. Furthermore we only want to resolve the code reference when the tab is actually loaded.
use `useExtensions` instead.

File: frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md (hunk 1 above)
Comment (jgbernalp) [reply frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md:987]: I added more descriptive docs. The `href` is the path appended to the existing URL, used elsewhere.

File: frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md (hunk 2 above)
Comment (jgbernalp) [reply frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md:980]: I changed and fixed the text, unfortunately when adding TSDoc `@example` the docs are not generated correctly. I checked the generation but it seems to be a task big enough for a different PR.

File: frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md (hunk 3 above)
Comment (jgbernalp) [reply frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md:986]: Added a list of possible values. Decided for horizontal nav in the docs.

File: frontend/public/components/utils/horizontal-nav.tsx (hunk 4 above)
Comment (jgbernalp) [reply frontend/public/components/utils/horizontal-nav.tsx:308]: I added a string based sorting for now. Until we include the `insertBefore` `insertAfter`

While checking the `insertBefore` `insertAfter` implementation seems other extensions are using  [order-extensions.ts](https://github.com/openshift/console/blob/master/frontend/packages/console-shared/src/utils/order-extensions.ts) which requires an id. Should we add the `id` property now, so adding the sorting later will be easier?

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/horizontal-nav-tabs.ts (hunk 5 above)
Comment (spadgett) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/horizontal-nav-tabs.ts:46]: I think we can deprecate page in the existing extension point and use a flatter structure, in which case we can flatten it here, too. We should make sure there is an issue tracking.

Comment (jgbernalp) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/horizontal-nav-tabs.ts]: @spadgett @christianvogt  Made the structure flat

File: frontend/public/components/utils/horizontal-nav.tsx [hunk 8]
```diff
@@ -276,22 +281,33 @@ export const HorizontalNav = React.memo((props: HorizontalNavProps) => {
...
+        ...tab.properties.page,
+        component: (params: PageComponentProps) => (
+          <AsyncComponent {...params} loader={tab.properties.component} />
+        ),
+      }));
+
+    const resolvedNavTab = dynamicTabExtensions
+      .filter((tab) => tab.properties.contextId === contextId)
+      .map((tab) => ({
+        ...tab.properties,
```
Comment (christianvogt) [inline frontend/public/components/utils/horizontal-nav.tsx:301]: You only need those properties that actually are relevant to be used later. ie `name` and `href`

**Comment by christianvogt** [comment]
minor comment but otherwise looks good
/approve

**Comment by jgbernalp** [comment]
/assign @yapei @opayne1 @RickJWagner
for approvals

**Comment by yapei** [comment]
with `console.tab` extension and all required property defined, user can add tab to any specific page matching with `contextId`
Screen Shot 2022-10-26 at 10 23 08
/label qe-approved

**Pull request merged**